test/unittests/util/test_date_util.py

```diff
@@ -0,0 +1,269 @@
+        with pytest.raises(BaseError):
+            parse_date_string(date_string)
+
+
+class TestSplitDateString:
+    """Tests splitting a date string into its components."""
+
+    def test_split_full_date_string(self) -> None:
+        date_string = "GREGORIAN:CE:2020-01-01:CE:2021-02-02"
+        result = _split_date_string(date_string)
+        expected = ("GREGORIAN", "CE", "2020-01-01", "CE", "2021-02-02")
+        assert result == expected
+
+    def test_split_start_date_only(self) -> None:
+        date_string = "2020-01-01"
+        result = _split_date_string(date_string)
+        expected = (None, None, "2020-01-01", None, None)
+        assert result == expected
+
+    def test_split_start_and_end_date_only(self) -> None:
+        date_string = "2020-01-01:2021-02-02"
+        result = _split_date_string(date_string)
+        expected = (None, None, "2020-01-01", None, "2021-02-02")
+        assert result == expected
+
+    def test_split_start_date_with_era(self) -> None:
+        date_string = "CE:2020-01-01"
+        result = _split_date_string(date_string)
+        expected = (None, "CE", "2020-01-01", None, None)
+        assert result == expected
+
+    def test_split_start_and_end_date_with_era(self) -> None:
+        date_string = "CE:2020-01-01:CE:2021-02-02"
+        result = _split_date_string(date_string)
+        expected = (None, "CE", "2020-01-01", "CE", "2021-02-02")
+        assert result == expected
+
+    def test_split_start_date_with_calendar(self) -> None:
+        date_string = "GREGORIAN:2020-01-01"
+        result = _split_date_string(date_string)
+        expected = ("GREGORIAN", None, "2020-01-01", None, None)
+        assert result == expected
+
+
+class TestParseSingleDate:
+    """Tests parsing a single date."""
+
+    def test_parse_date_string(self) -> None:
+        date_string = "2020-01-01"
+        result = _parse_single_date(date_string, Era.CE)
+        expected = SingleDate(era=Era.CE, year=2020, month=1, day=1)
+        assert result == expected
+
+    def test_parse_year_only(self) -> None:
+        date_string = "2020"
+        result = _parse_single_date(date_string, Era.CE)
+        expected = SingleDate(era=Era.CE, year=2020, month=None, day=None)
+        assert result == expected
+
+    def test_parse_year_and_month(self) -> None:
+        date_string = "2020-01"
+        result = _parse_single_date(date_string, Era.CE)
+        expected = SingleDate(era=Era.CE, year=2020, month=1, day=None)
+        assert result == expected
+
+    def test_parse_date_fewer_digits_for_month_and_day(self) -> None:
+        date_string = "0860-1-1"
+        result = _parse_single_date(date_string, Era.CE)
+        expected = SingleDate(era=Era.CE, year=860, month=1, day=1)
+        assert result == expected
+
+    def test_do_not_allow_fewer_digits_in_year(self) -> None:
+        date_string = "860-1-1"
+        with pytest.raises(BaseError):
+            _parse_single_date(date_string, Era.CE)
```

Comment:
not sure how I feel about that - maybe the regex should be changed to allow this